Finished DTOs (Activity and User), Finished PUT Endpoints

AleksandarRistic99 opened the pull request:
closes #64 
closes #23 
closes #66 
closes #75
closes #74 

- Added many tests in UserServiceTest, ControllerTest, UserServiceIntegrationTest, and DTO Mapper Tests
- pleas take a look at the Activity DTOs, to see if they match the requirements 
- added UserGetDTOProfile that gives public data to the frontend (UserGetDTO gives private data)
- UserPutDTO is for all the changes a user want to make on his profile/data
- added activityMapper
- take a look at the GetMapping getUserProfile in the UserController--> my idea is to return a UserGetDTO if a user is on own profile and to return a UserGetDTOProfile if user is on an other users profile
- added email and phone validation to handle invalid email and phone numbers before they reach database

NOTE: adding all the tests enabled a test coverage of 76% on the methods --> lets try to stay on that coverage to not have a stressful milestone 4 :)

![image](https://user-images.githubusercontent.com/71384880/115115835-dc029680-9f96-11eb-8c5c-da44f3eb5b9d.png)

AleksandarRistic99 committed:
StartedPutEndpoint

AleksandarRistic99 committed:
update

AleksandarRistic99 committed:
update

AleksandarRistic99 committed:
Removed GooglePOI Object

AleksandarRistic99 committed:
Update DTOMapper.java

AleksandarRistic99 committed:
Added ControllerTests

AleksandarRistic99 committed:
update

AleksandarRistic99 committed:
finished some tests

UserControllerTest, UserServiceInegrationTest,
UserServiceTest

AleksandarRistic99 committed:
Finished users DTOs

AleksandarRistic99 committed:
Finished User DTOs, added emailCheck

Also: Controllertests, Servicetests, ServiceIntegrationtests

AleksandarRistic99 committed:
Update UserServiceIntegrationTest.java

AleksandarRistic99 committed:
Update UserController.java

AleksandarRistic99 committed:
Added Attribute age

AleksandarRistic99 committed:
DTO Mappter test

AleksandarRistic99 committed:
Update DTOMapperTest.java

AleksandarRistic99 committed:
DTO MApper TEst

AleksandarRistic99 committed:
Finished DTOs for Activity and Tests

AleksandarRistic99 committed:
Added Phone number check and Email check

included tests

SamuelBruegger commented on src/main/java/ch/uzh/ifi/hase/soprafs21/rest/dto/userDTO/UserPutDTO.java:
```diff
@@ -0,0 +1,53 @@
+
+public class UserPutDTO {
+
+    private String email;
```
should the user be able to update his email?

AleksandarRistic99 replied:
Good point, i think they should have the possibility for cases like if they change university (uzh --> eth for example) and then to not be forced to register again, they could just change the email. But this is a point we can discuss on monday 👍

AleksandarRistic99 committed:
Update TokenValidation

Adapted token validation, such that invalid acces is not possible, if you want to access data from databse, you have to be a registered user

AleksandarRistic99 committed:
Small adaption

AleksandarRistic99 committed:
Added Password and Name validation and Tests

AleksandarRistic99 committed:
Additional Email check and Tests

AleksandarRistic99 committed:
Enabling Object Mapping

amoscalamida commented:
@AleksandarRistic99 Handle and Rethrow Validation Exceptions

<code>
try {
            // saves the given entity but data is only persisted in the database once flush() is called
            newUser = userRepository.save(newUser);
            userRepository.flush();
        }
        catch (ConstraintViolationException ex) { handleValidationError(ex); 
}

private void handleValidationError(ConstraintViolationException ex) {
        AtomicReference<String> exceptions = new AtomicReference<>();
        exceptions.set("Validation failed:\n");
        ex.getConstraintViolations().forEach(violation -> {
            exceptions.set(exceptions + violation.getMessage() + "\n");
        });
        throw new ResponseStatusException(HttpStatus.BAD_REQUEST, exceptions.toString());
    }
</code>

AleksandarRistic99 committed:
Adapting Email, Phone validation, removin ActivityPutDTO

Cleaning ActivityDTO and Tests

AleksandarRistic99 committed:
Merge branch 'master' into PUTEndpoint_ActivityDTO_UserDTO

Taremeh commented on src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UserController.java:
```diff
@@ -86,47 +89,69 @@ public void userPing(@PathVariable Long userId, @RequestHeader("Auth-Token")Stri
-    @ResponseStatus(HttpStatus.OK)
-    public void getUserProfile(@PathVariable Long userId, @RequestHeader("Auth-Token")String token){
+        // if user is on his own profile, he can see additional data like email --> return UserGetDTO
+        try{
+            if(userService.isUserAuthenticated(userId,token)){
+                return DTOMapperUser.INSTANCE.convertEntityToUserGetDTO(userFromRepo);
+            }
+        }catch(ResponseStatusException ignored){}
```
Why try & catch?

Taremeh replied:
Nevermind, it's b.c. of the auth function.
